enhancements/builds/volume-secrets.md

```diff
@@ -0,0 +1,129 @@
+---
+title: volume-mounted-secret-and-configmap-injection
+authors:
+  - "@bparees"
+reviewers:
+  - @adambkaplan
+  - @smarterclayton
+approvers:
+  - @adambkaplan
+creation-date: 2020-01-08
+last-updated: 2020-01-08
+status: provisional
+---
+
+# Volume Mounted Secret+ConfigMap Injections
+
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+ > 1. Can we make this the default or even only behavior for builds?  Probably not, need to make it opt-in to avoid potentially breaking existing buildconfigs.
+
+## Summary
+
+Today builds support getting source input from configmaps and secrets.  When users utilize this feature, the configmap or secret is volume-mounted into the build pod.  The next steps depend on whether it is an s2i or dockerfile build.
+
+For s2i builds, the generated Dockerfile contains commands to `ADD` the content at a path specified by the user, the assemble script is invoked, and then the injected content is zeroed out prior to committing the image via a `RUN rm` command added to the Dockerfile.
+
+For dockerfile builds, the user is instructed to add appropriate `ADD` and `RUN rm` commands to their dockerfile to inject the content that is available in the build's working directory (along with their application source, where applicable).
+
+There are a few undesirable aspects to this:
+1) In the dockerfile case, the content can still be found in lower layers of the image unless a layer squashing option is 
+selected.
+2) Requires extra work by the user in the Dockerfile, so each Dockerfile must be customized
+
+This enhancement proposes to use buildah's capability to mount a volume at build time.  The content mounted into the build pod would be then mounted into the container processing the Dockerfile, making that content available within the container so Dockerfile commands could reference it.  No explicit `ADD` would be required, and since mounted content is not committed to the resulting image, no `RUN rm` equivalent is required to clean up the injected content.
+
+
+## Motivation
+
+### Goals
+
+* Simplify how users consume secret + configmap content in builds
+* Increase the security of protected content being injected to images
+* Simplify use cases like mounting subscription entitlement credentials into a build so subscription content can be pulled down during a build while not including the credentials in the resulting image.
```

Comment:
an ancillary goal admittedly, but this feature (independent of the "build volumes" one noted below) could also be useful in tekton if/when we attempt to optimize integrating openshift builds as tekton tasks, as a way to facilitate input/output propagations between tasks.

So if that has any bearing in your eyes wrt prioritization, cite it as a positive reason for doing this.

Reply:
hm, can you elaborate on how we'd use it for that?

Reply:
I think I can:
1) the tekton pipeline define the config map / secret where data will be stored and then shared between tekton tasks
2) then within the same pipeline the various tasks are defined ... an initial task is defined that populates the config map / secret
3) maybe other tasks in the pipeline also use it
4) the openshift build based task then invokes our builder image with the necessary params / env vars that tells our build image to do the buildah mount in the way you described
5) and maybe tasks that run after the openshift build task also use the data

does that help ?

Reply:
yeah that helps.  that's basically possible today since we already support configmaps+secrets as "inputs" to the build.  This proposal just changes how the internals of us consuming that content works, for the most part.

So i'd say it's another use case that would be improved/simplified by this enhancement, but not a new use case that's enabled by it.

Reply:
that's a resonable interpretation of it ... so yeah, do with it what you will ... thanks